homeassistant/components/upnp/sensor.py

```diff
@@ -1,275 +1,257 @@
+        DerivedUpnpSensor(coordinator, device, SENSOR_TYPES[BYTES_RECEIVED]),
+        DerivedUpnpSensor(coordinator, device, SENSOR_TYPES[BYTES_SENT]),
+        DerivedUpnpSensor(coordinator, device, SENSOR_TYPES[PACKETS_RECEIVED]),
+        DerivedUpnpSensor(coordinator, device, SENSOR_TYPES[PACKETS_SENT]),
+    ]
+    async_add_entities(sensors, True)
 
 
 class UpnpSensor(Entity):
     """Base class for UPnP/IGD sensors."""
 
-    def __init__(self, device):
+    def __init__(
+        self,
+        coordinator: DataUpdateCoordinator,
+        device: Device,
+        sensor_type: Mapping[str, str],
+    ) -> None:
         """Initialize the base sensor."""
+        self._coordinator = coordinator
         self._device = device
+        self._sensor_type = sensor_type
 
-    async def async_added_to_hass(self):
-        """Subscribe to sensors events."""
-        self.async_on_remove(
-            async_dispatcher_connect(
-                self.hass, SIGNAL_REMOVE_SENSOR, self._upnp_remove_sensor
-            )
+    @property
+    def should_poll(self) -> bool:
+        """Inform we should not be polled."""
+        return False
+
+    @property
+    def icon(self) -> str:
+        """Icon to use in the frontend, if any."""
+        return "mdi:server-network"
+
+    @property
+    def available(self) -> bool:
+        """Return if entity is available."""
+        device_value_key = self._sensor_type["device_value_key"]
+        return (
+            self._coordinator.last_update_success
+            and device_value_key in self._coordinator.data
+            and not isinstance(self._coordinator.data[device_value_key], Exception)
         )
 
-    @callback
-    def _upnp_remove_sensor(self, device):
-        """Remove sensor."""
-        if self._device != device:
-            # not for us
-            return
+    @property
+    def name(self) -> str:
+        """Return the name of the sensor."""
+        return f"{self._device.name} {self._sensor_type['name']}"
 
-        self.hass.async_create_task(self.async_remove())
+    @property
+    def unique_id(self) -> str:
+        """Return an unique ID."""
+        return f"{self._device.udn}_{self._sensor_type['unique_id']}"
 
     @property
-    def device_info(self):
+    def unit_of_measurement(self) -> str:
+        """Return the unit of measurement of this entity, if any."""
+        return self._sensor_type["unit"]
+
+    @property
+    def device_info(self) -> Mapping[str, any]:
         """Get device info."""
         return {
             "connections": {(dr.CONNECTION_UPNP, self._device.udn)},
-            "identifiers": {(DOMAIN_UPNP, self._device.udn)},
             "name": self._device.name,
             "manufacturer": self._device.manufacturer,
             "model": self._device.model_name,
         }
 
+    async def async_added_to_hass(self) -> None:
+        """Subscribe to sensors events."""
+        self._coordinator.async_add_listener(self.async_write_ha_state)
```

Comment:
```suggestion
        self.async_on_remove(
            self._coordinator.async_add_listener(self.async_write_ha_state)
        )
```

Reply:
I think you don't meant that line 173 should be removed.